test/runtests.jl

```diff
@@ -80,6 +80,8 @@ include("setup_tests.jl")
                                 bytes2hex_(test_hash((1, (a=1, b=(x=1, y=2), c=(1, 2))))))
                 @test_reference("references/ref22_$(V)_$(nameof(hashfn)).txt",
                                 bytes2hex_(test_hash((;))))
+                @test_reference("references/ref23_$(V)_$(nameof(hashfn)).txt",
+                                ((; kwargs...) -> test_hash(kwargs))(; b=2, a=1))
```

Comment:
How does this test work? 

I see it references `references/ref23_$(V)_$(nameof(hashfn)).txt`, but I don't see a file with that name format I only see  `references/ref23.txt`

Also I see that it seems tests a NamedTuple IIUC, but this PR is focused on Pairs. Could you elaborate on that? I just want to make sure I'm not missing anything

Reply:
Oh, good catch. I forgot to commit the file!! 😬 

`@reference-test` writes the result to a file on the first run, and in future runs it checks that its result is the same as what is in the file. You can simply delete the file if you know the reference test should change. (But in practice, for `StableHashTraits`, if you expect a hash value to change, you should be bumping `V`).

Reply:
> Also I see that it seems tests a NamedTuple IIUC, but this PR is focused on Pairs. Could you elaborate on that? I just want to make sure I'm not missing anything

Note that we are passing `kwargs` value to `test_hash`. This is actually *not* a named tuple. It is a type of `Base.Pairs`. While this can be constructed into a named tuple (e.g. `NamedTuple(kwargs)`) it's actually a different type. In `StableHashTraits`, keyword arguments are hashed such that the keyword arguments are first sorted. This means that entering the keyword arguments in a different order doesn't change the hash value, which makes sense: for keyword arguments, it shouldn't matter what order you enter the arguments in, they mean the same thing. On the other hand `NamedTuple` objects generally have a meaningful order, and the hash changes depending on the order of the arguments.

Reply:
> Oh, good catch. I forgot to commit the file!! 😬
>
> `@reference-test` writes the result to a file on the first run, and in future runs it checks that its result is the same as what is in the file. 

Oh that's probably why it showed that tests all pass in CI when you had no files committed (which is weird behaviour IMO for CI). I wonder if there's a kwarg in the macro where if the file is non-existent it doesn't just silently pass the tests

Reply:
> Note that we are passing kwargs value to test_hash. This is actually not a named tuple. It is a type of Base.Pairs. While this can be constructed into a named tuple (e.g. NamedTuple(kwargs)) it's actually a different type. 

Oh I see what you mean now.

e.g.
```
julia> function foobar(; kwargs...)
           @info "kwarg" typeof(kwargs) kwargs
           return nothing
       end
foobar (generic function with 1 method)

julia> foobar(; b=2, a=1, c="kittens")
┌ Info: kwarg
│   typeof(kwargs) = Base.Pairs{Symbol, Any, Tuple{Symbol, Symbol, Symbol}, NamedTuple{(:b, :a, :c), Tuple{Int64, Int64, String}}}
│   kwargs =
│    pairs(::NamedTuple) with 3 entries:
│      :b => 2
│      :a => 1
└      :c => "kittens"
```

That said, is there a reason why you are using the kwargs interface in this PR? To me, if we wanted to test the stable hash behaviour on Pairs, couldn't do we do the following to test it directly: `test_hash(:A = 123)`?

But I imagine maybe you're maybe intending to hit two birds with one stone with this test, or possibly the main use-case is to make sure that kwargs are hashed correctly. Let me know your thoughts on that

Reply:
The goal is to directly test that `kwargs` hash correctly. 

Not entirely clear on your example, but note that `Pairs` is different from `Pair`s. 😅  (What's the `:` mean in your example, that looks like invalid syntax).

Reply:
Oh I see, I'll give this an approve! 

> Not entirely clear on your example, but note that Pairs is different from Pairs. 😅 (What's the : mean in your example, that looks like invalid syntax).

Ah sorry I had a typo in my example. I meant to type `=>`, i.e. I meant `:A => 123`, which would create a `Pair`. The `:` establishes a `Symbol`, e.g. see below

```
julia> :A => 123
:A => 123

julia> typeof(ans)
Pair{Symbol, Int64}
```

Reply:
Ah! With the `>` it makes sense now. Yeah, that would test `Pair`, which is separate from `Base.Pairs`.

Reply:
Also yes I was confusing `Pairs` vs `Pair` initially. Thanks for clarifying that!